File: typescript/2022/06.ts
// https://adventofcode.com/2022/day/6

function findUniqueCharacterSet(input: string, length: number): number {
  const recentUniqueChars: string[] = [];

  for (let i = 0; i < input.length; i++) {
    const char = input[i];
    const matchIndex = recentUniqueChars.indexOf(char);

    if (matchIndex !== -1) {
      // If this char is already in the list,
      // remove the chars up to the match and add the new char
      recentUniqueChars.splice(0, matchIndex + 1);
      recentUniqueChars.push(char);
    } else {
      // If not, just add the char to the list
      recentUniqueChars.push(char);

      if (recentUniqueChars.length === length) {
        return i + 1;
      }
    }
  }

  return -1;
}

function part1(input: string): number {
  return findUniqueCharacterSet(input, 4);
}

function part2(input: string): number {
  return findUniqueCharacterSet(input, 14);
}

const input = (await Bun.file("../_input/2022/06.txt").text()).trimEnd();

if (import.meta.env.NODE_ENV === "test") {
  const { test, expect } = await import("bun:test");

  test("part 1", () => expect(part1(input)).toBe(1198));
  test("part 2", () => expect(part2(input)).toBe(3120));
} else {
  console.log("part 1:", part1(input));
  console.log("part 2:", part2(input));
}

File: rust/2022/src/bin/06.rs
// https://adventofcode.com/2022/day/6

fn find_unique_set(input: &str, length: u16) -> u16 {
    let mut recent_unique_chars: Vec<char> = vec![];

    for (i, char) in input.chars().enumerate() {
        let char_position = recent_unique_chars.iter().position(|c| c == &char);

        match char_position {
            Some(position) => {
                // If this char is already in the list, remove
                // the chars up to the match and add the new char
                recent_unique_chars.drain(..position + 1);
                recent_unique_chars.push(char);
            }
            None => {
                // If not, just add the new char to the list
                recent_unique_chars.push(char);

                if recent_unique_chars.len() as u16 == length {
                    return i as u16 + 1;
                }
            }
        }
    }

    0
}

fn part1(input: &str) -> u16 {
    find_unique_set(input, 4)
}

fn part2(input: &str) -> u16 {
    find_unique_set(input, 14)
}

fn main() {
    let input = include_str!("../../../../_input/2022/06.txt").trim_end();

    println!("part 1: {}", part1(input));
    println!("part 2: {}", part2(input));
}

#[cfg(test)]
mod tests {
    const INPUT: &str = include_str!("../../../../_input/2022/06.txt");

    #[test]
    fn part1() {
        assert_eq!(super::part1(INPUT.trim_end()), 1198);
    }

    #[test]
    fn part2() {
        assert_eq!(super::part2(INPUT.trim_end()), 3120);
    }
}
